### PR TITLE
Confirmation rule prerequisite - fork choice filter change

### PR DESCRIPTION
**Pull request by saltiniroberto**
This PR specifies the changes to the Fork Choice specification required for the safe execution of the confirmation rule specified in #3339.

By implementing these changes, the confirmation rule can be safely run by means of beacon APIs calls, though there would be no connection to the execution layer.

See sections 3.1 to 3.4 of [this document](https://ethresear.ch/uploads/short-url/fV7zyTggJtMn8xlUUNVXTOB532G.pdf) for the rationale about the changes proposed and the related security analysis.

TODO:
- [x] Fix tests
- [x] Add changes to Bellatrix fork choice spec.

### CODE REVIEW

File: specs/phase0/fork-choice.md [hunk 1]
```diff
@@ -174,6 +175,13 @@ def get_current_slot(store: Store) -> Slot:
     return Slot(GENESIS_SLOT + get_slots_since_genesis(store))
 ```
 
+#### `get_current_store_epoch`
+
+```python
+def get_current_store_epoch(store: Store) -> Slot:
```
Comment (hwwhww) [inline specs/phase0/fork-choice.md:181]:
```suggestion
def get_current_store_epoch(store: Store) -> Epoch:
```

File: specs/phase0/fork-choice.md [hunk 2]
```diff
@@ -271,23 +287,17 @@ def filter_block_tree(store: Store, block_root: Root, blocks: Dict[Root, BeaconB
...
 
-    # If the previous epoch is justified, the block should be pulled-up. In this case, check that unrealized
-    # justification is higher than the store and that the voting source is not more than two epochs ago
-    if not correct_justified and is_previous_epoch_justified(store):
-        correct_justified = (
-            store.unrealized_justifications[block_root].epoch >= store.justified_checkpoint.epoch and
-            voting_source.epoch + 2 >= current_epoch
-        )
-
     finalized_checkpoint_block = get_checkpoint_block(
```
Comment (hwwhww) [inline specs/phase0/fork-choice.md:301]: It looks like a substantive modification that changed the test cases. Could you add some description about it in the PR subscription? 🙏

**Comment by mkalinin** [comment]
The change looks good to me. I don't think Bellatrix FC is affected by the change introduce in this PR. A question, would we need to add *new* tests along fixing the existing ones? (I guess no but asking just in case)

File: specs/phase0/fork-choice.md (hunk 2 above)
Comment (saltiniroberto) [reply specs/phase0/fork-choice.md:301]: Done

**Review by djrtwo — COMMENTED** [review]
I'm not sure I follow  -- is this to pull out a bit of functionality into specifically defined functions so that the beacon-api's can just reference these functions instead of trying to (re)define the behaviour there?

**Review by djrtwo — COMMENTED** [review]
I'm not sure I understand the connection to the Beacon-APIs here. can you elaborate?

**Comment by saltiniroberto** [comment]
> I'm not sure I understand the connection to the Beacon-APIs here. can you elaborate?

This PR includes the changes to the Fork Choice required to ensure the safety of the confirmation rule.

This means that, once this PR is implemented, the confirmation rule can be safely executed via standard Beacon API calls as done [here](https://gist.github.com/adiasg/4150de36181fd0f4b2351bef7b138893).

I have updated the title and PR description accordingly.

I hope that this clarifies.

**Review by djrtwo — COMMENTED** [review]
This looks good/correct.

Need to port to non-Phase0 functions as needed and then we can merge

**Comment by saltiniroberto** [comment]
> This looks good/correct.
> 
> Need to port to non-Phase0 functions as needed and then we can merge

Just checked and it looks like that no change is required to any of the non-Phase 0 specs.

**Comment by saltiniroberto** [comment]
We just need to confirm the test fix with @adiasg  and @mkalinin

**Comment by hwwhww** [comment]
The PR itself looks good to me. 👍 

But since it's a substantive change, we may need to be clear about when it can be on devnet/testnet/mainnet before merging it.

**Comment by etan-status** [comment]
- [v1.4.0-beta.2-hotfix+3431-consensus-spec-tests-mainnet.zip](https://github.com/ethereum/consensus-specs/files/12672917/v1.4.0-beta.2-hotfix%2B3431-consensus-spec-tests-mainnet.zip)
- [v1.4.0-beta.2-hotfix+3431-consensus-spec-tests-minimal.zip](https://github.com/ethereum/consensus-specs/files/12672921/v1.4.0-beta.2-hotfix%2B3431-consensus-spec-tests-minimal.zip)

**Review by mkalinin — APPROVED** [review]

**Review by djrtwo — APPROVED** [review]

**Review by hwwhww — APPROVED** [review]

**Pull request merged**